src/Cli/dotnet/SlnFileFactory.cs

```diff
@@ -19,87 +49,57 @@ public static string[] ListSolutionFilesInDirectory(string directory, bool inclu
 
-        public static SolutionModel CreateFromFileOrDirectory(string fileOrDirectory, bool includeSolutionXmlFiles = true)
+        public static SolutionModel CreateFromFileOrDirectory(string fileOrDirectory, bool includeSolutionFilterFiles = false, bool includeSolutionXmlFiles = true)
         {
-            if (File.Exists(fileOrDirectory))
-            {
-                return FromFile(fileOrDirectory);
-            }
-            else
-            {
-                return FromDirectory(fileOrDirectory, includeSolutionXmlFiles);
-            }
-        }
+            string solutionPath = GetSolutionFileFullPath(fileOrDirectory, includeSolutionFilterFiles, includeSolutionXmlFiles);
 
-        private static SolutionModel FromFile(string solutionPath)
-        {
-            SolutionModel slnFile = null;
-            try
+            if (solutionPath.HasExtension(".slnf"))
             {
-                ISolutionSerializer serializer = SolutionSerializers.GetSerializerByMoniker(solutionPath) ?? throw new GracefulException(
+                return CreateFromFilteredSolutionFile(solutionPath);
+            }
+            ISolutionSerializer serializer = SolutionSerializers.GetSerializerByMoniker(solutionPath) ?? throw new GracefulException(
                     CommonLocalizableStrings.CouldNotFindSolutionOrDirectory,
                     solutionPath);
 
-                slnFile = serializer.OpenAsync(solutionPath, CancellationToken.None).Result;
-            }
-            catch (SolutionException e)
-            {
-                throw new GracefulException(
-                    CommonLocalizableStrings.InvalidSolutionFormatString,
-                    solutionPath,
-                    e.Message);
-            }
-            return slnFile;
+            return serializer.OpenAsync(solutionPath, CancellationToken.None).Result;
         }
 
-        private static SolutionModel FromDirectory(string solutionDirectory, bool includeSolutionXmlFiles)
+        public static SolutionModel CreateFromFilteredSolutionFile(string filteredSolutionFile)
         {
-            DirectoryInfo dir;
-            try
-            {
-                dir = new DirectoryInfo(solutionDirectory);
-                if (!dir.Exists)
-                {
-                    throw new GracefulException(
-                        CommonLocalizableStrings.CouldNotFindSolutionOrDirectory,
-                        solutionDirectory);
-                }
-            }
-            catch (ArgumentException)
-            {
-                throw new GracefulException(
-                    CommonLocalizableStrings.CouldNotFindSolutionOrDirectory,
-                    solutionDirectory);
-            }
-
-            FileInfo[] files = [
-                ..dir.GetFiles("*.sln"),
-                ..(includeSolutionXmlFiles ? dir.GetFiles(".slnx") : [])
-             ];
+            JsonDocument jsonDocument = JsonDocument.Parse(File.ReadAllText(filteredSolutionFile));
```

Comment:
I'd rather we not manually parse the slnf is possible, as that provides more opportunities for bugs and more places that need to change if the format changes at all. Ideally, vs-solutionpersistence would handle this, but I understand that they don't yet. We could _consider_ using MSBuild's parser. It's currently internal, but they could theoretically make it public.

Note that if I were working on MSBuild right now, I'd advocate for not making it public because MSBuild does not and should not own the canonical parser for slnf, but since I don't, I will list it as a preferred option 😇

Reply:
Yes, I agree 100%. However, I reached out and told me they're a bit busy at the moment

Reply:
Also, I think the slnf grammar is simple enough as to run into few edge cases (and most places using the file would be msbuild anyways)